### PR TITLE
raftstore: 2nd attempt to fix deadlock when creating compaction guard

### PR DESCRIPTION
**Pull request by yiwu-arbug**
Signed-off-by: Yi Wu <yiwu@pingcap.com>

### What problem does this PR solve?

Issue Number: #9044

Problem Summary:
After the introduction of compaction guard (#8115), RocksDB compaction threads can call region info accessor to create SST partitioner, which blocks the compaction thread. After #8821 both of region info accessor and region tasks in moved into a unified background worker pool. Per #8990, the thread pool would contain 2-3 threads. The region tasks, such as `RegionTask::Apply` and `RegionTask::Destroy`, can execute file ingestions, which leads to a blocking flush in `crocksdb_ingest_external_file_optimized` in rust-rocksdb. It can create a deadlock when compaction is waiting in the background pool queue to create SST partitioner and multiple region tasks are waiting for RocksDB threads to run flush.

Since the region tasks contain blocking flush call, which could run for a while, it doesn't sound like a good idea to have them run in the background pool, which could block other background pool tasks. Moving region info access to its own worker is a precaution to avoid similar deadlock happen in the future.

### What is changed and how it works?

What's Changed: 
Move region info accessor and region tasks each out of the unified background worker pool to their own worker thread. 

Also in compaction_guard.rs, access region info provider in `should_partition()` instead of `create_partitioner()`, since the later can be invoked by RocksDB while holding db_mutex.

### Related changes

compaction guard was added in #8115. 
SST cleanup task worker is merged with background worker in #8821. 
The first attempt to fix the deadlock was #9045.
background thread pool size was default to 2 instead of 1 in #8990, which should relieve the problem, but apparently it is not.

This PR depends on https://github.com/tikv/rust-rocksdb/pull/598.

### Check List 

Tests 

Will do manual test.

### Release note 

- Fix deadlock when creating compaction guard.

### CODE REVIEW

**Comment by yiwu-arbug** [comment]
WIP for pending test.

**Comment by yiwu-arbug** [comment]
@Little-Wallace @Connor1996 any suggestion on how can I implement an unit test?

File: components/raftstore/src/coprocessor/region_info_accessor.rs [hunk 1]
```diff
@@ -490,23 +490,31 @@ impl RunnableWithTimer for RegionCollector {
...
     scheduler: Scheduler<RegionInfoQuery>,
 }
 
 impl RegionInfoAccessor {
     /// Creates a new `RegionInfoAccessor` and register to `host`.
     /// `RegionInfoAccessor` doesn't need, and should not be created more than once. If it's needed
     /// in different places, just clone it, and their contents are shared.
-    pub fn new(host: &mut CoprocessorHost<impl KvEngine>, worker: &Worker) -> Self {
+    pub fn new(host: &mut CoprocessorHost<impl KvEngine>) -> Self {
+        let worker = WorkerBuilder::new("region-info").create();
```
Comment (Connor1996) [inline components/raftstore/src/coprocessor/region_info_accessor.rs:507]: name it "region-collector-worker" too?

**Comment by Connor1996** [comment]
> @Little-Wallace @Connor1996 any suggestion on how can I implement an unit test?

Schedule destroy-region task and do a manual compact?

**Comment by yiwu-arbug** [comment]
> > @Little-Wallace @Connor1996 any suggestion on how can I implement an unit test?
> 
> Schedule destroy-region task and do a manual compact?

Should it be at raftstore level or at server level? Since the issue involve the server level `background_worker`.

**Comment by yiwu-arbug** [comment]
I'm to move region accessor query from `SstPartitionerFactory::create_partitioner` to `SstPartitioner::should_partition`, since rocksdb can call `create_partitioner` while holding db_mutex, which can block other TiKV operations that read/write rocksdb.

File: Cargo.toml [hunk 2]
```diff
@@ -218,6 +218,9 @@ raft-proto = { git = "https://github.com/tikv/raft-rs", branch = "master", defau
 protobuf = { git = "https://github.com/pingcap/rust-protobuf", rev = "65e9df20fbcbcf2409d5ee86a2332ecd04c534f8" }
 protobuf-codegen = { git = "https://github.com/pingcap/rust-protobuf", rev = "65e9df20fbcbcf2409d5ee86a2332ecd04c534f8" }
 
+[patch.'https://github.com/tikv/rust-rocksdb']
+rocksdb = { git = "https://github.com/yiwu-arbug/rust-rocksdb", branch = "guard_mut" }
```
Comment (Connor1996) [inline Cargo.toml:222]: Please update this

**Review by Connor1996 — APPROVED** [review]
rest LGTM

**Comment by yiwu-arbug** [comment]
> > > > @Little-Wallace @Connor1996 any suggestion on how can I implement an unit test?
> > > 
> > > 
> > > Schedule destroy-region task and do a manual compact?
> > 
> > 
> > Should it be at raftstore level or at server level? Since the issue involve the server level `background_worker`.
> 
> Maybe in integrations/raftstore?

will update the unit test in a separate PR.

File: components/raftstore/src/store/compaction_guard.rs [hunk 3]
```diff
@@ -48,60 +48,87 @@ impl<P: RegionInfoProvider> CompactionGuardGeneratorFactory<P> {
...
-
-impl SstPartitioner for CompactionGuardGenerator {
-    fn should_partition(&self, req: &SstPartitionerRequest) -> SstPartitionerResult {
-        let mut pos = self.pos.get();
+impl<P: RegionInfoProvider> SstPartitioner for CompactionGuardGenerator<P> {
+    fn should_partition(&mut self, req: &SstPartitionerRequest) -> SstPartitionerResult {
+        if !self.initialized {
+            self.initialize();
+        }
+        if !self.use_guard {
```
Comment (Little-Wallace) [inline components/raftstore/src/store/compaction_guard.rs:128]: Should we check `use_guard` before `initialize `?

**Review by Little-Wallace — APPROVED** [review]
LGTM

**Pull request merged**